fs.go

```diff
@@ -0,0 +1,197 @@
+package dalec
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"path"
+	"sync"
+
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/tonistiigi/fsutil/types"
+
+	"github.com/tonistiigi/fsutil"
+
+	"github.com/moby/buildkit/client/llb"
+)
+
+var _ fs.DirEntry = &stateRefDirEntry{}
+var _ fs.ReadDirFS = &StateRefFS{}
+
+type StateRefFS struct {
+	s       llb.State
+	ctx     context.Context
+	opts    []llb.ConstraintsOpt
+	client  gwclient.Client
+	res     *gwclient.Result
+	o       sync.Once
+	initErr error
+}
+
+func NewStateRefFS(s llb.State, ctx context.Context, client gwclient.Client) *StateRefFS {
```

Comment:
Can you move this into the `frontend` package/
In general I try to keep everything that requires the gateway client and other such types out of the main dalec package.